### PR TITLE
Constructor selection based on @inject annotation, reformulated binding registration and circular dependency verification

### PR DESCRIPTION
**Pull request by Dreckr**
## Constructor selection based on @inject

The constructor is selected following this rules:

1- Is there any constructor annotated with @inject?
2- Is there only one constructor?
3- Is there a no-args constructor?

The first rule to pass is used to define the constructor. In case none of this rules pass, an exception is raised.
## Reformulated binding registration

Instead of creating provider functions for everything and having a map for singletons, now there is an actual Binding abstract class that defines a interface so that the instance construction and scope management logic can be abstracted from the Injector.
The main benefits of this new binding registration on this pull request is that constructors only have to be selected once for each binding (_ConstructorBinding stores the selected constructor so it can always use it without having to search for it again) and provider bindings can be verified for circular dependencies (_ProviderBinding stores the MethodMirror so it can know its parameters).
##  Circular dependency verification

After all the bindings were registered, they are analyzed to see if there is any circular dependency. This verification DOES NOT work in cases where the injector itself is injected (this includes bindTo cases, as the injector is injected silently into the module, as the constructor or provider can try to inject itself using it and there is no way to see this on the injectors initialization.

### CODE REVIEW

**Comment by Dreckr** [comment]
Closed for now, as more changes are coming

File: lib/dado.dart [hunk 1]
```diff
@@ -59,11 +59,21 @@ library dado;
...
 import 'src/mirror_utils.dart';
 
-typedef Object _Provider(Injector injector);
-
-Symbol _typeName(Type type) => reflectClass(type).qualifiedName;
+part 'src/binding.dart';
+
+Symbol _typeName(type) {
+  if (type is Type)
+    return reflectClass(type).qualifiedName;
```
Comment (justinfagnani) [inline lib/dado.dart]: style: put if statement blocks in braces unless they fit on one line

**Comment by justinfagnani** [comment]
Awesome Diego, I'll take a look this weekend.

BTW, I hate reviews on Github. I'll try it out, but would you mind too much if we uploaded this to Reitveld so we can get side-by-side reviews? I does require installing Google's depot tools, but it integrates nicely with local git repos.

**Comment by Dreckr** [comment]
Ok, I've never used it before, but I'll give it a try.

**Comment by justinfagnani** [comment]
Thanks! Let me know if you have any trouble. The instructions are here: https://code.google.com/p/dart/wiki/PreparingYourMachine

**Comment by Dreckr** [comment]
Here it is: https://codereview.appspot.com/11927045/

Only upload.py is needed though...

**Comment by justinfagnani** [comment]
Manually merged in https://github.com/dart-lang/dado/commit/cdb06e60dc1d649a6cacea1f01b2005dce3bc56c

Thanks Diego!!

**Comment by Dreckr** [comment]
Awesome! I'm sorry I made this take so long, I'll make the next pull requests easier to merge.

Could you create issues for every feature and change you would like to see in Dado? I will have some spare time and I want to contribute a little more.

**Pull request closed without merging**